module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,730 @@
+    fn slash_and_invalidate_attestation(
+        &self,
+        attester: &C::Address,
+        transition_nb: u64,
+        reason: SlashingReason,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        let reward = self.slash_user_helper(attester, Role::Attester, working_set);
+
+        let curr_reward_value = self
+            .bad_transition_pool
+            .get(&transition_nb, working_set)
+            .unwrap_or_default();
+
+        self.bad_transition_pool
+            .set(&transition_nb, &(curr_reward_value + reward), working_set);
+
+        Err(AttesterIncentiveErrors::UserSlashed(reason))
+    }
+
+    /// A helper function for the `bond_challenger/attester` call. Also used to bond challengers/attesters
+    /// during genesis when no context is available.
+    pub(super) fn bond_user_helper(
+        &self,
+        bond_amount: u64,
+        user_address: &C::Address,
+        role: Role,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<CallResponse> {
+        // Transfer the bond amount from the sender to the module's address.
+        // On failure, no state is changed
+        let coins = Coins {
+            token_address: self
+                .bonding_token_address
+                .get(working_set)
+                .expect("Bonding token address must be set"),
+            amount: bond_amount,
+        };
+        self.bank
+            .transfer_from(user_address, &self.address, coins, working_set)?;
+
+        let (balances, event_key) = match role {
+            Role::Attester => (&self.bonded_attesters, "bonded_attester"),
+            Role::Challenger => (&self.bonded_challengers, "bonded_challenger"),
+        };
+
+        // Update our record of the total bonded amount for the sender.
+        // This update is infallible, so no value can be destroyed.
+        let old_balance = balances.get(user_address, working_set).unwrap_or_default();
+        let total_balance = old_balance + bond_amount;
+        balances.set(user_address, &total_balance, working_set);
+
+        // Emit the bonding event
+        working_set.add_event(
+            event_key,
+            &format!("new_deposit: {bond_amount:?}. total_bond: {total_balance:?}"),
+        );
+
+        Ok(CallResponse::default())
+    }
+
+    /// Try to unbond the requested amount of coins with context.sender() as the beneficiary.
+    pub(crate) fn unbond_challenger(
+        &self,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse> {
+        // Get the user's old balance.
+        if let Some(old_balance) = self.bonded_challengers.get(context.sender(), working_set) {
+            // Transfer the bond amount from the sender to the module's address.
+            // On failure, no state is changed
+            self.reward_sender(old_balance, context, working_set)?;
```

Comment:
We follow the convention that `context` is always the first argument to a method (after `self`)

Reply:
Done